src/ocean_data_assim/MOM_oda_incupd.F90

```diff
@@ -0,0 +1,874 @@
+  type(thermo_var_ptrs),     intent(in)    :: tv !< A structure pointing to various thermodynamic variables
+
+  real, target, dimension(SZIB_(G),SZJ_(G),SZK_(GV)), &
+                           intent(in)      :: u    !< The zonal velocity that is being
+                                                   !! initialized [L T-1 ~> m s-1]
+  real, target, dimension(SZI_(G),SZJB_(G),SZK_(GV)), &
+                           intent(in)      :: v    !< The meridional velocity that is being
+                                                   !! initialized [L T-1 ~> m s-1]
+  type(oda_incupd_CS),       pointer       :: CS !< A pointer to the control structure for this module
+                                                 !! that is set by a previous call to initialize_oda_incupd (in).
+
+
+  real, dimension(SZK_(GV)) :: tmp_val1          ! data values on the model grid
+  real, allocatable, dimension(:) :: tmp_val2    ! data values remapped to increment grid
+  real, allocatable, dimension(:,:,:) :: h_obs   !< h of increments
+  real, allocatable, dimension(:) :: tmp_h       ! temporary array for corrected h_obs
+  real, allocatable, dimension(:) :: hu_obs,hv_obs  ! A column of thicknesses at h points [H ~> m or kg m-2]
+  real, dimension(SZK_(GV)) :: hu, hv            ! A column of thicknesses at h, u or v points [H ~> m or kg m-2]
+
+
+  integer ::  i, j, k, is, ie, js, je, nz, nz_data
+  integer :: isB, ieB, jsB, jeB
+  real :: h_neglect, h_neglect_edge
+  real :: sum_h1, sum_h2 !vertical sums of h's
+  character(len=256) :: mesg
+
+  is = G%isc ; ie = G%iec ; js = G%jsc ; je = G%jec ; nz = GV%ke
+  isB = G%iscB ; ieB = G%iecB ; jsB = G%jscB ; jeB = G%jecB
+  if (.not.associated(CS)) return
+
+
+  ! increments calculated on if CS%ncount = 0.0
+  if (CS%ncount /= 0.0) call MOM_error(FATAL,'calc_oda_increments: '// &
+           'CS%ncount should be 0.0 to get accurate increments.')
+
+
+  if (GV%Boussinesq) then
+    h_neglect = GV%m_to_H*1.0e-30 ; h_neglect_edge = GV%m_to_H*1.0e-10
+  else
+    h_neglect = GV%H_subroundoff ; h_neglect_edge = GV%H_subroundoff
+  endif
+
+  ! get h_obs
+  nz_data = CS%Inc(1)%nz_data
+  allocate(h_obs(G%isd:G%ied,G%jsd:G%jed,nz_data)) ; h_obs(:,:,:) = 0.0
+  do k=1,nz_data  ; do j=js,je ; do i=is,ie
+    h_obs(i,j,k) = CS%Ref_h%p(i,j,k)
+  enddo ; enddo ; enddo
+  call pass_var(h_obs,G%Domain)
+
+
+  ! allocate 1-d arrays
+  allocate(tmp_h(nz_data)); tmp_h(:) = 0.0
+  allocate(tmp_val2(nz_data)) ; tmp_val2(:) = 0.0
+  allocate(hu_obs(nz_data)) ; hu_obs(:) = 0.0
+  allocate(hv_obs(nz_data)) ; hv_obs(:) = 0.0
+
+  ! remap t,s (on h_init) to h_obs to get increment
+  tmp_val1(:) = 0.0
+  do j=js,je ; do i=is,ie
+     if (G%mask2dT(i,j) == 1) then
+        ! account for the different SSH
+        sum_h1 = 0.0
+        sum_h2 = 0.0
+        do k=1,nz
+          sum_h1 = sum_h1+h(i,j,k)
+        enddo
+
+        do k=1,nz_data
+          sum_h2 = sum_h2+h_obs(i,j,k)
+          tmp_h(k)=(sum_h1/sum_h2)*h_obs(i,j,k)
```

Comment:
Actually, I need to rescind my approval ... sum_h2 should be in its own loop since the final result is used by tmp_h

Reply:
@abozec : do you expect any answer change for the change in sum_h2 loop ?

Reply:
This calculation is done only when you are using full analysis fields (instead of increments) as input and in the current state, the increment calculated from the full field is wrong.  If you are already using increments as input, the answer does not change.

Reply:
@abozec good catch,  i will make these changes and test.